test/common/buffer/watermark_buffer_test.cc

```diff
@@ -62,6 +62,58 @@ TEST_F(WatermarkBufferTest, AddBuffer) {
   EXPECT_EQ(11, buffer_.length());
 }
 
+TEST_F(WatermarkBufferTest, Prepend) {
+  std::string suffix = "World!", prefix = "Hello, ";
+
+  buffer_.add(suffix.data(), suffix.size());
+  EXPECT_EQ(0, times_high_watermark_called_);
+  buffer_.prepend(prefix.data(), prefix.size());
+  EXPECT_EQ(1, times_high_watermark_called_);
+  EXPECT_EQ(suffix.size() + prefix.size(), buffer_.length());
+}
+
+TEST_F(WatermarkBufferTest, PrependToEmptyBuffer) {
+  std::string suffix = "World!", prefix = "Hello, ";
+
+  buffer_.prepend(suffix);
+  EXPECT_EQ(0, times_high_watermark_called_);
+  EXPECT_EQ(suffix.size(), buffer_.length());
+
+  buffer_.prepend(prefix.data(), prefix.size());
+  EXPECT_EQ(1, times_high_watermark_called_);
+  EXPECT_EQ(suffix.size() + prefix.size(), buffer_.length());
+
+  buffer_.prepend("");
+  EXPECT_EQ(1, times_high_watermark_called_);
+  EXPECT_EQ(suffix.size() + prefix.size(), buffer_.length());
+}
+
+TEST_F(WatermarkBufferTest, PrependString) {
+  std::string suffix = "World!", prefix = "Hello, ";
+
+  buffer_.add(suffix);
+  EXPECT_EQ(0, times_high_watermark_called_);
+  buffer_.prepend(prefix);
+  EXPECT_EQ(1, times_high_watermark_called_);
+  EXPECT_EQ(suffix.size() + prefix.size(), buffer_.length());
+}
+
+TEST_F(WatermarkBufferTest, PrependBuffer) {
+  std::string suffix = "World!", prefix = "Hello, ";
+  buffer_.add(suffix);
+  EXPECT_EQ(0, times_high_watermark_called_);
+  EXPECT_EQ(suffix.size(), buffer_.length());
+
+  OwnedImpl prefixBuffer;
+  prefixBuffer.add(prefix);
+  buffer_.prepend(prefixBuffer);
+
+  EXPECT_EQ(1, times_high_watermark_called_);
```

Comment:
Oops, I failed to ask for the regression test I wanted (though I wanted these too!).  Can we also verify if we move data from buffer A to buffer B using prepend, that if A goes from over high watermark to under low watermark?    Basically we want to have a regression test that fails if we remove that postProcess you added.

I think we can just changed the buffer in this test to a WatermarkBuffer with small limits, assert when we add the initial data it goes high, and when data is removed via prepend-drain that it goes back low.

Reply:
Yep! I thought of that test case right after I pushed. Will update soon.